doc/source/numpy_2_0_migration_guide.rst

```diff
@@ -2,13 +2,48 @@
 NumPy 2.0 Migration Guide
 *************************
 
-This document contains a set of instructions on how to update your code to work with
-the Numpy 2.0 Python API. Most of the changes are trivial, and require the end user
-to use a different name/module to access a given function/constant.
+This document contains a set of instructions on how to update your code to
+work with Numpy 2.0.
+
+
+.. _migration_windows_int64:
+
+Windows default integer
+=======================
+The default integer used by NumPy is now 64bit on all 64bit systems (and
+32bit on 32bit system).  For historic reasons related to Python 2 it was
+previously equivalent to the C ``long`` type.
+The default integer is now equivalent to ``np.intp``.
+
+Most end-users should not be affected by this change.  Some operations will
+use more memory, but some operations may actually become faster.
+If you experience issues due to calling a library written in a compiled
+language it may help to explicitly cast to a ``long``, for example with:
+``arr = arr.astype("long", copy=False)``.
+
+Libraries interfacing with compiled code that are written in C, Cython, or
+a similar language may require updating to accomodate user input if they
+are using the ``long`` or equivalent type on the C-side.
+In this case, you may wish to using ``intp`` and cast user input or support
+both ``long`` and ``intp`` (to better support NumPy 1.x as well).
+When creating a new integer array in C or Cython, the new ``NPY_DEFAULT_INT``
```

Comment:
Maybe we should define this in 1.26 too?

Reply:
Might make it a bit easier to use for downstream, but they still will need to force `numpy>=1.26` at compile time.  So I am tempted to say that unfortunately, it is better to vendor it:
```
#ifndef NPY_DEFAULT_INT
#define NPY_DEFAULT_INT NPY_LONG
#endif
```
(unless you force `>2.0` at compile time)?

Happy to follow-up with a backport, not sure it's helpful, but wouldn't hurt.